### PR TITLE
[EDA] Lead Conversion Account Naming Setting

### PR DESCRIPTION
**Pull request by mrrigney**
# Changes
## Updates for Lead Conversion
We added a new EDA Setting called Automatically Rename Lead-Converted Accounts. Now, admins can choose to automatically rename Accounts created when a Lead is converted to use the account naming format that's associated with its record type. Without this setting, the Account name is determined by the Company field on the Lead record. 

For example, let’s say you selected the Administrative record type in Automatically Rename Lead-Converted Accounts. When you convert a Lead and specify the Administrative record type, the Account record is renamed to use the Administrative Account Name Format from EDA Settings. An Account for a Lead named Louis Lombard is automatically renamed Lombard Administrative Account.

Starting with the trial version of EDA 1.xxx, this setting is configured to specify the Administrative and Household Account record types by default. If you started with a version of EDA earlier than 1.xxx or you're starting with EDA 1.xxx via the installer rather than the trial, no default record types are specified.

For more information about working with Leads, see [Lead](https://powerofus.force.com/s/article/EDA-Lead) in the EDA Documentation. 

# Issues Closed
[[W-9020537]](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000008xk4tIAA/view)

# New Metadata
## Custom Labels
- stgLeadConversionAccountNamingTitle
- stgLeadConversionAccountNaming
- stgLeadConversionAccountNamingDesc

## Apex
- AccountModelSettingsController
- LeadConversionAccountNamingMapper
- LeadConversionAccountNamingMapper_TEST
- LeadConversionAccountNamingModel
- LeadConversionAccountNamingModel_TEST
- LeadConversionAccountNamingService
- LeadConversionAccountNamingService_TEST
- LeadConversionAccountNamingVMapper
- LeadConversionAccountNamingVMapper_TEST

# Testing Notes

* Verify when the EDA Settings come up (New EDA Settings QA Only), the Lead Conversion Auto Account Naming comes up after Rename Household Accounts
* Add new Account Record Types and verify they show up in the Available Record Types
* Select 1 or more Record Types and verify they move over to the Selected Account Types
* Click Save, refresh the page, and verify the Selected Record Types are still there
* Remove 1 or more Record Types from the Selected Account Record Types
* Click Save, refresh the page, and verify the Records are Removed and moved to the Available Account Record Types
* Go to Setup/Custom Settings, and Select Hierarchy Settings
* Verify the Lead Naming Account Record Types field is populated with the Selected Record Types
* Verify all of the existing functionality within Account Model are still working as expected, pay particular attention to the Accounts Without Contacts Section.

### CODE REVIEW

**Comment by paulmalenke** [comment]
@mrrigney I added @agoolsby and me to the PR as reviewers for visibility.

File: force-app/main/default/classes/AccountModelSettingsController.cls [hunk 1]
```diff
@@ -64,6 +64,22 @@ public with sharing class AccountModelSettingsController {
         }
     }
 
+    /************************************************************************************
+     * @description Retrieve a DualListboxVModel represeting the current Account Record Types
```
Comment (paulmalenke) [inline force-app/main/default/classes/AccountModelSettingsController.cls:68]:
```suggestion
     * @description Retrieve a DualListboxVModel representing the current Account Record Types
```

File: force-app/main/default/classes/LeadConversionAccountNamingMapper.cls [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        }
+
+        return instance;
+    }
+
+    /*****************************************************************************
+    * @description Retrieves an LeadConversionAccountNamingModel populated with
+    * a list of Account Record Type Ids specified in Hierarchy Settings 
+    * selected for renaming Accounts during Lead Conversion.
+    * @return A1 LeadConversionAccountNamingModel
```
Comment (paulmalenke) [inline force-app/main/default/classes/LeadConversionAccountNamingMapper.cls:63]:
```suggestion
    * @return A LeadConversionAccountNamingModel
```

File: force-app/main/default/classes/LeadConversionAccountNamingVMapper.cls [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            instance = new LeadConversionAccountNamingVMapper();
+        }
+
+        return instance;
+    }
+
+    /*******************************************************************************************************
+     * @description Retrieve a DualListboxVModel representing the Account Record Types currently selected
+     * for renaming accounts during Lead Conversion
+     * availabe for selection.
```
Comment (paulmalenke) [inline force-app/main/default/classes/LeadConversionAccountNamingVMapper.cls:63]:
```suggestion
     * @description Retrieve a DualListboxVModel representing the Account Record Types currently selected
     * for renaming accounts during Lead Conversion that available for selection.
```

File: force-app/main/default/lwc/accountModelSettings/accountModelSettings.js [hunk 4]
```diff
@@ -73,6 +80,15 @@ export default class AccountModelSettings extends LightningElement {
         }
     }
 
+    @wire(getLeadConversionAccountNamingSettingsViewModel)
+    leadConversionAccountNamingSettingsViewModel({ error, data }) {
+        if (data) {
+            this.leadConversionAccountNamingSettingsViewModel = data;
+        } else if (error) {
+            //console.log("error retrieving leadConversionAccountNamingSettingsViewModel");
```
Comment (paulmalenke) [inline force-app/main/default/lwc/accountModelSettings/accountModelSettings.js:88]: I am not sure if this was intentional or not. I will leave it up to EDA reviewers to provide insight into EDA UI error handling.

**Review by paulmalenke — CHANGES_REQUESTED** [review]
Looks good to me. I did make some minor suggestions and was able to test and verify the functionality. I will rely on the EDA reviewers to go more in-depth around style and etc.

File: force-app/main/default/lwc/edaSettingsNavigation/edaSettingsNavigation.js [hunk 5]
```diff
@@ -37,6 +37,7 @@ export default class EdaSettingsNavigation extends LightningElement {
             adminAccountNaming: "Administrative Account Name Format",
             hhAccountNaming: "Household Account Name Format",
             autoHHAccountNaming: "Automatically Rename Household Accounts",
+            leadConversionAutoAccountNaming: "Lead Conversion Auto Account Naming",
```
Comment (MatthewBlanski) [inline force-app/main/default/lwc/edaSettingsNavigation/edaSettingsNavigation.js:40]: Please have this use stgLeadConversionAccountNaming and import it in the label imports above.

File: force-app/main/default/lwc/accountModelSettings/accountModelSettings.js [hunk 6]
```diff
@@ -192,15 +212,11 @@ export default class AccountModelSettings extends LightningElement {
...
-        // add updated setting to hierarchySettingsChanges object
+    handleLeadNamingAccountChange(event) {
+        // add selected values to hierarchySettingsChanges object
         let hierarchySettingsChange = {
-            settingsType: "string",
-            settingsName: "Admin_Account_Naming_Format__c",
+            settingsType: "array",
+            settingsName: "Lead_Naming_Account_RTypes__c",
             settingsValue: event.detail.value,
         };
```
Comment (MatthewBlanski) [inline force-app/main/default/lwc/accountModelSettings/accountModelSettings.js:221]: Please restore the Administrative Account Naming Change method and have the lead naming account change as a separate method, as it seems was your original intent.

File: force-app/main/default/classes/LeadConversionAccountNamingMapper_TEST.cls [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+        LeadConversionAccountNamingMapper LeadConversionAccountNamingMapperInstance = 
+                LeadConversionAccountNamingMapper.getInstance();
+
+        Test.startTest();
+        LeadConversionAccountNamingModel LeadConversionAccountNamingModel = 
+            LeadConversionAccountNamingMapperInstance.getLeadConversionAccountNamingModel();
+        Test.stopTest();
+
+        System.debug('Orphaned list ' + LeadConversionAccountNamingModel.accountRecordTypeIdsForLeadConversionWhenOrphanedList);
```
Comment (MatthewBlanski) [inline force-app/main/default/classes/LeadConversionAccountNamingMapper_TEST.cls:115]: Please remove this debug.
```suggestion
```

**Review by MatthewBlanski — CHANGES_REQUESTED** [review]
This is functional overall and matches EDA's style. There are a few minor details and a method that looks like it got swept away accidentally in the case of Administrative Account Naming. A minor tweak to bring that back, removal of a system debug statement in a test, and the addition of a label in the nav will finish this one out.

File: force-app/main/default/lwc/edaSettingsNavigation/edaSettingsNavigation.js (hunk 5 above)
Comment (mrrigney) [reply force-app/main/default/lwc/edaSettingsNavigation/edaSettingsNavigation.js:40]: Modified to use the label

File: force-app/main/default/lwc/accountModelSettings/accountModelSettings.js (hunk 4 above)
Comment (mrrigney) [reply force-app/main/default/lwc/accountModelSettings/accountModelSettings.js]: @MatthewBlanski just confirming this is OK with you, I was following the pattern of the other @wire's

**Comment by MatthewBlanski** [comment]
This is acceptable. We don't have UI for displaying the errors yet, so
we're leaving it be for now.

Matthew Blanski
Lead Member of Technical Staff
Salesforce - Salesforce.org
C: 773-326-7278

File: force-app/main/default/lwc/accountModelSettings/accountModelSettings.js (hunk 6 above)
Comment (mrrigney) [reply force-app/main/default/lwc/accountModelSettings/accountModelSettings.js]: Added handleAdministrativeAccountNamingChange back

**Comment by mrrigney** [comment]
@MatthewBlanski There is a line in the spec about defaulting the Lead Account Naming values for new orgs.  Can you point me in the direction to add these defaults and what about existing orgs?
"For new orgs, the Administrative and Household Account record types will be added to the list of selected record types by default. For existing orgs, the customer must select the record types they want to use (nothing will be preselected)."

**Comment by MatthewBlanski** [comment]
> @MatthewBlanski There is a line in the spec about defaulting the Lead Account Naming values for new orgs. Can you point me in the direction to add these defaults and what about existing orgs?
> "For new orgs, the Administrative and Household Account record types will be added to the list of selected record types by default. For existing orgs, the customer must select the record types they want to use (nothing will be preselected)."

That would be in UTIL_CustomSettingsFacade.configSettings()

**Review by MatthewBlanski — APPROVED** [review]

**Review by paulmalenke — APPROVED** [review]

File: force-app/main/default/labels/CustomLabels.labels-meta.xml [hunk 8]
```diff
@@ -2109,7 +2109,31 @@
...
         <shortDescription>Description of the Accounts Without Contacts setting</shortDescription>
         <value>Delete Accounts with these record types if they have no Contacts.</value>
-    </labels>
+    </labels>  
+    <labels>
+        <fullName>stgLeadConversionAccountNamingDesc</fullName>
+        <categories>Settings</categories>
+        <language>en_US</language>
+        <protected>true</protected>
+        <shortDescription>Description of the Auto Account Naming setting</shortDescription>
```
Comment (agoolsby) [inline force-app/main/default/labels/CustomLabels.labels-meta.xml:2118]:
```suggestion
        <shortDescription>Description of the Automatically Rename Lead-Converted Accounts setting</shortDescription>
```

File: force-app/main/default/objects/Hierarchy_Settings__c/fields/Lead_Naming_Account_RTypes__c.field-meta.xml [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<CustomField xmlns="http://soap.sforce.com/2006/04/metadata">
+    <fullName>Lead_Naming_Account_RTypes__c</fullName>
```
Comment (agoolsby) [inline force-app/main/default/objects/Hierarchy_Settings__c/fields/Lead_Naming_Account_RTypes__c.field-meta.xml:3]:
```suggestion
    <fullName>Lead_Converted_Account_RTypes__c</fullName>
```

File: force-app/main/default/objects/Hierarchy_Settings__c/fields/Lead_Naming_Account_RTypes__c.field-meta.xml [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+    <description>The record types of Accounts to use EDA Account Naming for during Lead Conversion.</description>
```
Comment (agoolsby) [inline force-app/main/default/objects/Hierarchy_Settings__c/fields/Lead_Naming_Account_RTypes__c.field-meta.xml:4]:
```suggestion
    <description>The Account record types to use EDA&apos;s account naming format when converting Leads.</description>
```

File: force-app/main/default/objects/Hierarchy_Settings__c/fields/Lead_Naming_Account_RTypes__c.field-meta.xml [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+    <externalId>false</externalId>
+    <inlineHelpText>The record types of Accounts to use EDA Account Naming for during Lead Conversion.</inlineHelpText>
```
Comment (agoolsby) [inline force-app/main/default/objects/Hierarchy_Settings__c/fields/Lead_Naming_Account_RTypes__c.field-meta.xml:6]:
```suggestion
    <inlineHelpText>The Account record types to use EDA&apos;s account naming format when converting Leads.</inlineHelpText>
```

File: force-app/main/default/objects/Hierarchy_Settings__c/fields/Lead_Naming_Account_RTypes__c.field-meta.xml [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+    <label>Lead Naming Account Record Types</label>
```
Comment (agoolsby) [inline force-app/main/default/objects/Hierarchy_Settings__c/fields/Lead_Naming_Account_RTypes__c.field-meta.xml:7]:
```suggestion
    <label>Lead-Converted Account Record Types</label>
```

**Review by agoolsby — CHANGES_REQUESTED** [review]
Left suggestions for the final label changes. Thanks @mrrigney. FYI @juliesnow

**Comment by agoolsby** [comment]
@juliesnow I've updated the Release Notes for this PR. Please review and make any adjustments as you see fit to align with how other EDA RNs are treated. 

FYI @mrrigney You might want to take a look at the RN too. I know that this PR does not cover the "trigger" that powers this magic, but I think this description sets the stage. Let me know what you think. Thanks!

File: force-app/main/default/objects/Hierarchy_Settings__c/fields/Lead_Naming_Account_RTypes__c.field-meta.xml (hunk 9 above)
Comment (mrrigney) [reply force-app/main/default/objects/Hierarchy_Settings__c/fields/Lead_Naming_Account_RTypes__c.field-meta.xml:3]: I'm not committing this suggestion directly from here as it requires quite a few other changes, but have made the change.

File: force-app/main/default/labels/CustomLabels.labels-meta.xml (hunk 8 above)
Comment (agoolsby) [reply force-app/main/default/labels/CustomLabels.labels-meta.xml:2118]: @mrrigney  Looks like this suggestion got resolved without being committed.

**Review by agoolsby — CHANGES_REQUESTED** [review]
Just one suggestion that didn't get committed.

**Review by agoolsby — APPROVED** [review]

**Review by annejackson — APPROVED** [review]
Approved by me! @spinkelman feel free to merge when you are done!

**Review by spinkelman — APPROVED** [review]
Approved. I see no issues and no conflicts in existing EDA functionality. New labels have been called out in our LQA for translation with 232. @annejackson

**Comment by spinkelman** [comment]
Reviews completed. Merging. Will notify all teams once all builds pass.

**Pull request merged**